### ISSUE TITLE
[Feature Request] Support for nested tensors

### ISSUE THREAD

**Issue by vmoens** [pytorch/tensordict#99]
## Motivation

Support for nested tensors should work as `torch.stack(tensordicts, 0)`
(we can't currently override `torch.nested.as_nested_tensor` so we'll need to write a custom op)

We don't need to subclass TensorDict for this but there are a few caveats:
1. Reshaping won't be allowed
2. Shape can't be accessed
3. Indexing can only be done on the first dim

For instance, populating a tensordict with a nested tensor won't work as we can't access the shape.

- If nested tensors are found in a tensordict, all tensors should have that feature.

- Indexing such a tensordict along the second dimension will require:
1. splitting the tensors that are nested
2. indexing those tensors
3. re-nesting them

**Comment by zihaozou** [comment]
any update on this?

**Comment by vmoens** [comment]
Yes there's a bunch of things we can do, and some we can't:
```python
from tensordict import TensorDict
import torch

# You can place nested tensors in TD
td = TensorDict(a=torch.nested.nested_tensor([
    torch.arange(10),
    torch.arange(9),
    torch.arange(8),
], layout=torch.jagged), batch_size=[3])

# you can index TD with an int
td[0]

# you can unbind
td.unbind(0)

# You can consolidate in a single storage
td.consolidate()

# This breaks bc TensorDict relies on foreach for add - we should have a workaround
td = td + 1 # breaks bc foreach not implemented

# you can't index a nested tensor with a slice atm
td[:1] # breaks
```

Another interesting thing is that you can turn a lazy stack into a td with nested tensors:
```python
from tensordict import TensorDict, lazy_stack
import torch

td0 = TensorDict(a=torch.arange(10))
td1 = TensorDict(a=torch.arange(9))
td = lazy_stack([td0, td1])
assert td[0] is td0
tdstack = td.densify(layout=torch.jagged)
assert tdstack[0] is not td0
print(tdstack)
```
which prints
```
TensorDict(
    fields={
        a: NestedTensor(shape=torch.Size([2, j1]), device=cpu, dtype=torch.int64, is_shared=False)},
    batch_size=torch.Size([2]),
    device=None,
    is_shared=False)
```

There's plenty of opportunity of stuff to do! If there's any use case that is broken for you, please submit an issue an we'll look into it!

cc @jbschlosser
